File: java-workshop/part-2-light-mocks/src/test/java/service/ShippingServiceTest.java
package service;

import domain.Order;
import domain.OrderItem;
import domain.OrderStatus;
import domain.ShippingConfirmation;
import gateway.ShippingGateway;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.*;

class ShippingServiceTest {

    @Test
    void createShipment_callsGatewayAndReturnsConfirmation() {
        ShippingGateway shippingGateway = mock(ShippingGateway.class);
        GatewayShippingService shippingService = new GatewayShippingService(shippingGateway);

        var order = new Order(
            "ORDER-123",
            "CUST-456",
            List.of(
                new OrderItem("PROD-001", 2, 10.0),
                new OrderItem("PROD-002", 1, 20.0)
            ),
            OrderStatus.PENDING,
            0.0,
            null,
            null
        );
        var expectedConfirmation = new ShippingConfirmation("ORDER-123", "TRACK-ABC123", "2024-12-25");

        when(shippingGateway.createShipment(any())).thenReturn(expectedConfirmation);

        ShippingConfirmation result = shippingService.createShipment(order);

        assertEquals(expectedConfirmation, result);
        verify(shippingGateway).createShipment(argThat(request ->
            request.orderId().equals("ORDER-123")
                && request.customerId().equals("CUST-456")
                && request.items().size() == 2
        ));
    }

    @Test
    void cancelShipment_callsGatewayAndReturnsResult() {
        ShippingGateway shippingGateway = mock(ShippingGateway.class);
        GatewayShippingService shippingService = new GatewayShippingService(shippingGateway);

        var order = new Order(
            "ORDER-123",
            "CUST-456",
            List.of(
                new OrderItem("PROD-001", 2, 10.0),
                new OrderItem("PROD-002", 1, 20.0)
            ),
            OrderStatus.PENDING,
            0.0,
            "TXN-789",
            "TRACK-ABC123"
        );

        when(shippingGateway.createShipment(any())).thenReturn(
            new ShippingConfirmation(order.id(), order.trackingNumber(), "2024-12-25")
        );
        when(shippingGateway.cancelShipment("TRACK-ABC123")).thenReturn(true);

        assertTrue(shippingService.cancelShipment("TRACK-ABC123"));
        verify(shippingGateway).cancelShipment("TRACK-ABC123");
    }
}

File: kotlin-workshop/part-2-light-mocks/src/test/kotlin/service/ShippingServiceTest.kt
package service

import domain.*
import gateway.ShippingGateway
import io.mockk.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ShippingServiceTest {

    private lateinit var shippingGateway: ShippingGateway
    private lateinit var shippingService: ShippingService

    @BeforeEach
    fun setUp() {
        shippingGateway = mockk()
        shippingService = ShippingService(shippingGateway)
    }

    @Test
    fun `createShipment calls gateway and returns confirmation`() {
        // Given
        val order = Order(
            id = "ORDER-123",
            customerId = "CUST-456",
            items = listOf(
                OrderItem("PROD-001", 2, 10.0),
                OrderItem("PROD-002", 1, 20.0)
            )
        )

        val expectedConfirmation = ShippingConfirmation(
            orderId = "ORDER-123",
            trackingNumber = "TRACK-ABC123",
            estimatedDelivery = "2024-12-25"
        )

        every { shippingGateway.createShipment(any()) } returns expectedConfirmation

        // When
        val result = shippingService.createShipment(order)

        // Then
        assertEquals(expectedConfirmation, result)
        verify {
            shippingGateway.createShipment(match { request ->
                request.orderId == "ORDER-123" &&
                request.customerId == "CUST-456" &&
                request.items.size == 2
            })
        }
    }

    @Test
    fun `cancelShipment returns true when gateway succeeds`() {
        // Given
        every { shippingGateway.cancelShipment("TRACK-ABC123") } returns true

        // When
        val result = shippingService.cancelShipment("TRACK-ABC123")

        // Then
        assertTrue(result)
        verify(exactly = 1) { shippingGateway.cancelShipment("TRACK-ABC123") }
    }

    @Test
    fun `cancelShipment returns false when gateway fails`() {
        // Given
        every { shippingGateway.cancelShipment("TRACK-UNKNOWN") } returns false

        // When
        val result = shippingService.cancelShipment("TRACK-UNKNOWN")

        // Then
        assertFalse(result)
        verify(exactly = 1) { shippingGateway.cancelShipment("TRACK-UNKNOWN") }
    }
}

